### ISSUE TITLE
jax.jnp gives NaN values

### ISSUE THREAD

**Issue by sajastu** [google/jax#12185]
### Description

I'm trying to calculate average evaluation loss of ~39K values, but it seems that the summation operation produces NaN value. Here is the snippet that causes this:

eval_metrics = []
```
for i, batch_idx in enumerate(tqdm(eval_batch_idx, desc="Evaluating ...", position=2)):
	# Model forward
    metrics = pad_shard_unpad(p_eval_step, static_return=True)(
        state.params, model_inputs.data, min_device_batch=per_device_eval_batch_size
    )
    # jax.debug.print('Loss_2: {}', metrics['loss'])
    # jax.debug.print('normalizer: {}', metrics['normalizer'])
    eval_metrics.append(metrics)

eval_metrics = get_metrics(eval_metrics)
eval_metrics = jax.tree_map(jnp.sum, eval_metrics)
eval_metrics = jax.tree_map(lambda x: x / eval_normalizer, eval_metrics)
```
`get_metrics(eval_metrics)` returns arrrays of values (39K values), and `jax.tree_map(jnp.sum, eval_metrics)` sums these values (the root of the problem), producing NaN value for the arrays whose summation overflows. 

### What jax/jaxlib version are you using?

jax 0.3.16

### Which accelerator(s) are you using?

GPU

### Additional System Info

Linux Ubuntu 18.04

**Comment by jakevdp** [comment]
It's difficult to understand what's happening without more information. Can you put together a complete minimal example? (i.e. including data and imports of the functions you're using?)

**Comment by sajastu** [comment]
Hi @jakevdp 

The posted snippet is actually a part of the Flax on[ pertaining BART language model with Huggingface](https://github.com/huggingface/transformers/blob/main/examples/flax/language-modeling/run_bart_dlm_flax.py). I essentially try to use that code on my custom dataset, whose validation instances are about ~39K, so the returning jnp array has 39K values that should be normalized to display. 

Actually I did a further debugging using `jax.debug.print()`. Here is the losses within the returned list: 

eval_metrics["loss"]= `[22339.582 22161.816 22221.883 ... 22303.713 23481.447       nan]`

So it seems that the NaN is caused by a specific instance, that makes the who summation return NaN. 

I debugged the program to find this instance, and noticed that one row of the input_ids within the eval_step function is essentially all zero, that makes the returned logits NaN. Here are the input_ids as well as the computed logits:

```
input_ids [[50264   179  4783 ...     1     1     1]
 [ 9226   118 50264 ...     1     1     1]
 [  700 50264  1794 ...     1     1     1]
 [    0     0     0 ...     0     0     0]] 
 logits [[[ -8.235256    -2.0338707   -1.2944565  ...  -7.842334    -8.178107
    -8.643117  ]
  [ -5.334233    -3.0293756   -0.13147672 ... -10.615038   -10.16274
    -9.98302   ]
  [ -6.7125335   -2.661169     0.44605067 ... -10.435076   -10.56175
    -8.552453  ]
  ...
  [ -4.102307    -1.8319362    3.401112   ...  -9.376341    -9.284803
    -6.1490345 ]
  [ -7.457472    -3.5980809    1.4830123  ... -11.469392   -11.51775
   -10.788591  ]
  [ -5.4814677   -2.5836527    1.274877   ... -11.137861   -11.14509
   -11.144929  ]]

 [[ -8.236677    -2.0341225   -1.2935951  ...  -7.842687    -8.178189
    -8.642431  ]
  [ -6.3445244   -2.9560397   -0.50635713 ... -10.960117   -11.018831
   -10.380886  ]
  [ -4.027665    -1.2919675    1.4371854  ...  -7.0959826   -6.739063
    -6.0127587 ]
  ...
  [ -4.9435964   -2.7261846    0.84278244 ... -13.782334   -14.252498
   -13.15108   ]
  [ -2.3252742   -3.1369848   -0.84928894 ... -13.953517   -14.834746
   -14.477739  ]
  [ -4.905314    -4.5334086   -0.95402426 ... -11.740114   -12.343864
   -11.699981  ]]

 [[ -8.233315    -2.03387     -1.2937447  ...  -7.843014    -8.178542
    -8.64241   ]
  [ -8.00472     -2.0967667   -0.24173795 ... -10.587084   -10.671728
    -9.482687  ]
  [ -3.4650757   -2.165937    -0.75387245 ... -11.202366   -11.33529
   -10.740507  ]
  ...
  [ -4.254065    -2.3328125    0.44079754 ... -11.385041   -11.237292
    -9.75622   ]
  [ -6.5205693   -1.8131874   -0.44322762 ... -13.063074   -13.735182
   -13.360048  ]
  [ -4.8410645   -2.7622364   -0.9499095  ... -14.075225   -14.371651
   -14.25375   ]]

 [[         nan          nan          nan ...          nan          nan
            nan]
  [         nan          nan          nan ...          nan          nan
            nan]
  [         nan          nan          nan ...          nan          nan
            nan]
  ...
  [         nan          nan          nan ...          nan          nan
            nan]
  [         nan          nan          nan ...          nan          nan
            nan]
  [         nan          nan          nan ...          nan          nan
            nan]]]
```


Moreover, I checked the input values before calling the `eval_step` function, and here are those related to the above example:
{

```
'input_ids': array([[  463,   118,   119, ...,     1,     1,     1],
       [15010, 50264,   560, ...,     1,     1,     1],
       [  118,   298,  3252, ...,     1,     1,     1],
       ...,
       [10669, 12465, 50264, ...,     1,     1,     1],
       [25800,   627, 50264, ...,     1,     1,     1],
       [ 8877, 50264,  6025, ...,     1,     1,     1]]), 

'labels': array([[  605,   995,  1250, ...,  1794,   560, 26628],
       [    4,  -100,  4297, ...,  9902, 18028,   354],
       [  844,     6,  1409, ..., 17137,    43,     6],
       ...,
       [17137,  2463,   281, ..., 24638, 41404, 29676],
       [  118,  8494, 36024, ..., 30035, 35507,  4783],
       [ 4484,   560,  1610, ...,  7443, 11970,  4297]]),

 'decoder_input_ids': array([[    2,   605,   995, ...,  1990,  1794,   560],
       [    2,     4,     1, ...,  4783,  9902, 18028],
       [    2,   844,     6, ...,   428, 17137,    43],
       ...,
       [    2, 17137,  2463, ...,  3654, 24638, 41404],
       [    2,   118,  8494, ...,  7333, 30035, 35507],
       [    2,  4484,   560, ...,  1459,  7443, 11970]]),

 'attention_mask': array([[1, 1, 1, ..., 0, 0, 0],
       [1, 1, 1, ..., 0, 0, 0],
       [1, 1, 1, ..., 0, 0, 0],
       ...,
       [1, 1, 1, ..., 0, 0, 0],
       [1, 1, 1, ..., 0, 0, 0],
       [1, 1, 1, ..., 0, 0, 0]]),

 'decoder_attention_mask': array([[1, 1, 1, ..., 1, 1, 1],
       [1, 1, 0, ..., 1, 1, 1],
       [1, 1, 1, ..., 1, 1, 1],
       ...,
       [1, 1, 1, ..., 1, 1, 1],
       [1, 1, 1, ..., 1, 1, 1],
       [1, 1, 1, ..., 1, 1, 1]])

}
```

I'm stuck at this stage; wondering how the Jax could modify the input values? Since there's no row whose all values is 0, but in Jax function, we do see that there's such a row.

**Comment by sajastu** [comment]
UPDATE:

Through debugging the posted code, I could finally find the root of the problem! It seems that JAX pads the given input with 0's given the batch_size in the experiments. However, the BART model treats 0 as a special token. The error comes from the inconsistency between the JAX padding value (0) and BART's padding value (1) as the input_ids. As a workaround for it, I made my inputs all fit in batch_size. For instance, if the last batch includes 7 instances (like the one above) which is less than 8, I just drop that sample. Given my evaluation set size, I don't have any problem with dropping the last batch of data, but for other experiments (where we want to keep and evaluate all instances within the evaluation set), I believe we need to change padding value of the JAX and make it consistent with the padding value that the core model (Here, BART) takes in. 

The source snippet uses `flax.jax_utils.pad_shard_unpad(...)` to call the JAX utilities, but I'm not sure how we can change the padding value from 0 to 1 in this function. Probably, we need to call more specific functions [as instructed here](https://flax.readthedocs.io/en/latest/guides/full_eval.html#manual-implementation).

**Comment by jakevdp** [comment]
Thanks for the update – it sounds like JAX is behaving as expected. Perhaps it's worth opening a flax issue?
